### PR TITLE
Cleanup Antag Selection Logic a Lot

### LINKED ISSUE #40606: Fixtures don't work on entities spawned by gamerules until said entity moves
### What version did the issue occur in?

last commit on the master branch: 27f281add5a608c3a90abbbba218fb63075278fa

### Description

When a gamerule spawns entities. they have fixtures turned off until they move. they can still be attacked via melee. but projectiles don't work. 

Spawning entities via the F5 menu doesn't cause this issue.

Image

### Reproduction

1. Add a gamerule like Nukeops
2. Take a ghostrole (but once you get it, DON'T MOVE)
3. Aghost and try to shoot it with any gun

### Additional Context

For some reason this bug didn't happen with a reinforcement monkey the nukeops rule spawned

### PR DESCRIPTION
**Pull request by Princess-Cheeseballs**
## About the PR

I got really annoyed an antag selection system today and rather than let it go like a normal person I let it fester in my mind all day preventing me from doing anything productive until I had improved it at least a little bit.

Fixes #40606

Likely qualifies as a Mald PR

## Why / Balance

Bugfixes and cleanup.

The system was too generic for its own good and had no safeguards to prevent spawning entities in nullspace for example and just leaving them there. Wouldn't even throw an error or warn if it did.

In addition, the behavior of spawning and initializing an entity in nullspace, then moving it to the correct grid resulted in physics fixture bugs that were bad. It was this issue specifically that brought me to refactoring some of this system because I had just made an engine PR that fixed that bug along with some others.

In addition, spawning required that any whitelists apply to both the spawner AND entity being spawned or else it would spawn you at the wrong location. I shouldn't need to describe how this was limiting and stupid and easily prone to bugs.

## Technical details

Basically I looked at every possible execution path and split them up into separate methods then combined them like legos. I then removed anything that wasn't being used to simplify stuff since there were a lot of handlers for "but what if this thing isn't here" when it 100% should be there. 

To test, turn yourself into an antag with verbs, gamerules, and ghost role spawners. Use the antag verbs as a ghost as well.
They should all work.

I also made a proper const string for a station testing map, and set it to Saltern since that's what the map leads said was the current load bearing station map. This also makes it easier to change our load bearing station map in the future since it's one const string. 

## Media

https://github.com/user-attachments/assets/90725197-5b02-4691-ad59-ef19955ea9fa

## Requirements

- [X] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [X] I have added media to this PR or it does not require an in-game showcase.

## Breaking changes

MakeAntag() has been deleted entirely. Please use ForceMakeAntag<T> or TryMakeAntag instead. 

**Changelog**

### CODE REVIEW

**Comment by Princess-Cheeseballs** [comment]
Test fail real and correctly caught by one of the errors I added. Ebic.

**Comment by Princess-Cheeseballs** [comment]
Should've fixed the test fail, the issue was paradox clone needs a station to spawn but the test only created empty space meaning paradox clone was unable to spawn.

File: Content.Server/Antag/AntagSelectionSystem.cs [hunk 1]
```diff
@@ -38,6 +38,11 @@
 
 namespace Content.Server.Antag;
 
+/// <summary>
+/// Turns players into antags.
+/// </summary>
+/// <remarks>DO NOT EVER EVER EVER SPAWN A NEW ENTITY IN NULLSPACE THEN MOVE IT TO THE GRID.
```
Comment (beck-thompson) [inline Content.Server/Antag/AntagSelectionSystem.cs]: The reminder is good for sure but I personally dont all caps comments like this 😆
maybe tone it down a bit?

File: Content.Server/Antag/AntagSelectionSystem.cs [hunk 2]
```diff
@@ -353,70 +352,47 @@ public void AssignPreSelectedSessions(Entity<AntagSelectionComponent> ent)
...
 
     /// <summary>
-    /// Makes a given player into the specified antagonist.
+    /// Create an antag spawner which can be taken over by a player through the ghost role system.
     /// </summary>
-    public void MakeAntag(Entity<AntagSelectionComponent> ent, ICommonSession? session, AntagSelectionDefinition def, bool ignoreSpawner = false)
+    /// <param name="ent">Antag rule entity</param>
+    /// <param name="def">Antag selection definition chosen from the entity</param>
+    /// <param name="session">Player session if it exists</param>
+    private void CreateAntagSpawner(Entity<AntagSelectionComponent> ent, AntagSelectionDefinition def, ICommonSession? session = null)
```
Comment (beck-thompson) [inline Content.Server/Antag/AntagSelectionSystem.cs]: remove the session! It looks like its unused now

File: Content.Server/Antag/AntagSelectionSystem.cs [hunk 3]
```diff
@@ -353,70 +352,47 @@ public void AssignPreSelectedSessions(Entity<AntagSelectionComponent> ent)
...
+            Del(spawner);
         }
 
-        if (antagEnt is not { } player)
+        if (!TryComp<GhostRoleAntagSpawnerComponent>(spawner, out var spawnerComp))
         {
-            Log.Error($"Attempted to make {session} antagonist in gamerule {ToPrettyString(ent)} but there was no valid entity for player.");
-            _adminLogger.Add(LogType.AntagSelection, $"Attempted to make {session} antagonist in gamerule {ToPrettyString(ent)} but there was no valid entity for player.");
-            if (session != null && ent.Comp.RemoveUponFailedSpawn)
+            Log.Error($"Antag spawner {spawner} does not have a GhostRoleAntagSpawnerComponent.");
```
Comment (Tayrtahn) [inline Content.Server/Antag/AntagSelectionSystem.cs]:
```suggestion
            Log.Error($"Antag spawner {spawner} does not have a {nameof(GhostRoleAntagSpawnerComponent)}.");
```
Using `nameof` for stuff like this makes it easier to rename the class in the future (since IDEs can auto-replace instances in code.

File: Content.Server/Antag/AntagSelectionSystem.cs [hunk 4]
```diff
@@ -353,70 +352,47 @@ public void AssignPreSelectedSessions(Entity<AntagSelectionComponent> ent)
...
+            _adminLogger.Add(LogType.AntagSelection, $"Antag spawner {spawner} in gamerule {ToPrettyString(ent)} failed due to not having GhostRoleAntagSpawnerComponent.");
```
Comment (Tayrtahn) [inline Content.Server/Antag/AntagSelectionSystem.cs]:
```suggestion
            _adminLogger.Add(LogType.AntagSelection, $"Antag spawner {spawner} in gamerule {ToPrettyString(ent)} failed due to not having {nameof(GhostRoleAntagSpawnerComponent)}.");
```

File: Content.Server/Antag/AntagSelectionSystem.cs (hunk 1 above)
Comment (Princess-Cheeseballs) [reply Content.Server/Antag/AntagSelectionSystem.cs]: It's not the only place in the code base that suffers from this issue unfortunately. So I think it's important to remind people. But I can remove the all caps if it's really egregious.

Comment (beck-thompson) [reply Content.Server/Antag/AntagSelectionSystem.cs]: > It's not the only place in the code base that suffers from this issue unfortunately. So I think it's important to remind people. But I can remove the all caps if it's really egregious.

Yeah like I said the reminder is def good and should stay, I personally think its important to try and keep comments professional (even though it might be a valid crashout :laughing:!)

File: Content.Server/Antag/AntagSelectionSystem.cs [hunk 5]
```diff
@@ -38,6 +38,12 @@
 
 namespace Content.Server.Antag;
 
+/// <summary>
+/// Turns players into antags.
+/// </summary>
+/// Do not ever ever ever spawn and initialize an entity prototype in nullspace then move it to the grid.
+/// I wasted 4 hours refactoring this system specifically to fix that mistake.
+/// Always initialize your entities attached to the entity you're spawning them on, or the correct map at the very least.
```
Comment (Tayrtahn) [inline Content.Server/Antag/AntagSelectionSystem.cs:48]:
```suggestion
/// <remarks>
/// Do not ever ever ever spawn and initialize an entity prototype in nullspace then move it to the grid.
/// I wasted 4 hours refactoring this system specifically to fix that mistake.
/// Always initialize your entities attached to the entity you're spawning them on, or the correct map at the very least.
/// </remarks>
```

File: Content.Server/Antag/AntagSelectionSystem.cs
Comment (Tayrtahn) [inline Content.Server/Antag/AntagSelectionSystem.cs:1]: I feel like most of these methods should return the spawned entity (or have it as an out parameter). I guess you can get it from the passed session's attached entity, but that seems a little clumsy.

**Review by Tayrtahn — APPROVED** [review]

**Comment by ThatGuyUSA** [comment]
Closes #42673 ?

**Comment by Princess-Cheeseballs** [comment]
> Closes #42673 ?
> 
> 

Unfortunately no.

**Comment by Princess-Cheeseballs** [comment]
Actually nvm all the issues I raised were covered by this PR. 
System still needs more refactoring imo but yeah that issue can be closed

**Comment by ThatGuyUSA** [comment]
>System still needs more refactoring imo

I would appreciate opening a new issue about refactoring, I only closed it because you fixed all of the issues you mentioned.
Spreading awareness is good!
